### PR TITLE
Optics based on kotlin.reflect

### PR DESCRIPTION
**Pull request by serras**
The issue #2418 (by @magneticflux-) has given me an idea for an API for optics based on `KProperty`. It's a bit rough on the edges (for example, there's no way to check at compile-time that you are referring to a field in a `data class`, so the access may fail at runtime). On the other hand, it allows you to create optics easily for types you do not control:

```kotlin
data class Person(val name: String, val age: String, val friends: List<Person>)

val friendNames = Person::friends.every compose Person::name.lens
friendNames.modify(someFriends) { it.capitalize() }
```

The prism part is done using a combinator `instance<Parent, Children>`. Maybe a better name is needed, though.

### CODE REVIEW

File: arrow-libs/optics/arrow-optics/build.gradle.kts
Comment (nomisRev) [inline arrow-libs/optics/arrow-optics/build.gradle.kts:29]: 😱

**Comment by nomisRev** [comment]
As a rule Arrow never exports unsafe functions from the main APIs.
I was hoping to solve this issue with a compiler plugin, but they've been on hold for a long time since Optics was released.
Soon they might become stable though, and creating optics for classes you don't own will be automatically solved as a part of that.

With that in mind, that'd also mean we'd have to deprecate this.
We could try looking into doing this with [Reflekt](https://github.com/JetBrains-Research/reflekt) though.

**Comment by nomisRev** [comment]
Or perhaps in a second module `arrow-optics-reflect` so it can be opted-in.

**Comment by serras** [comment]
> Or perhaps in a second module `arrow-optics-reflect` so it can be opted-in.

I can do that, sure. The goal here was to provide a simple bridge which works for now.

**Comment by nomisRev** [comment]
Yes, that makes perfect sense. We've historically done that in other projects too, there is a small reflection based module in Arrow Endpoints for `Schema` too but hopefully we can replace all those with nicer plugins next year 🥳

Thanks for adding this! When moved to `arrow-optics-reflect`, and perhaps a small addition in the docs we can merge 👍

**Comment by serras** [comment]
@nomisRev I've just pushed a commit which moves this to a new `arrow-optics-reflect` package.

File: arrow-libs/optics/arrow-optics-reflect/build.gradle.kts
Comment (nomisRev) [inline arrow-libs/optics/arrow-optics-reflect/build.gradle.kts:19]: So this exposes the API of Kotlin Std, and uses the JDK8 implementation?

I'm still not a 100% sure how this works, and what is the correct config :/ 
I don't we're doing different things in different modules.. but off-topic for this PR.

**Review by nomisRev — COMMENTED** [review]
Yes, that looks perfect @serras! Thanks for moving it 👍
Some docs would be great too ❤️

Let me know when you think this is ready, and I'll review and ✔️

File: arrow-libs/optics/arrow-optics-reflect/build.gradle.kts
Comment (serras) [reply arrow-libs/optics/arrow-optics-reflect/build.gradle.kts]: I'm not sure 100% either. In any case, I've removed the `api` dependency on `stdlibCommon` and just left the `implementation` for the JDK version.

**Comment by nomisRev** [comment]
@serras should become green after merging `main`.

**Review by nomisRev — APPROVED** [review]

**Review by raulraja — APPROVED** [review]

**Pull request merged**
